### ISSUE TITLE
Cannot restart task after failure

### ISSUE THREAD

**Issue by moricard** [mesosphere/marathon-lb#177]
I run marathon-lb docker container in privileged mode and host network as a marathon task to handle service discovery and load balancing for my services. I recently have seen some unexpected behaviour.

For no apparent reason, the health check starts failing and the marathon-lb task stops working normally. Once the task has failed it will never start properly on the same machine, it has to be started on a new slave that has never seen this task fail. The container logs show normal behaviour but the task is unresponsive. 

What am I missing here?

docker version 1.10.3
marathon version 0.15.3-1.0.463.ubuntu1404
mesos version 0.28.0-2.0.16.ubuntu1404

**Comment by brndnmtthws** [comment]
Can you check the iptables rules? I'm wondering if they aren't being removed correctly after a reload. Run:

```
# iptables --list
```

Also, is it possible that haproxy stopped running for any reason? Perhaps there is an error in the config due to some custom templates?

**Comment by moricard** [comment]
Thanks for the quick reply, the haproxy runs correctly but when running `iptables --list`, I still see rules for the marathon-lb stuff:

```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
DROP       tcp  --  anywhere             anywhere             tcp dpt:9090 flags:FIN,SYN,RST,ACK/SYN

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination
DOCKER-ISOLATION  all  --  anywhere             anywhere
DOCKER     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain DOCKER (1 references)
target     prot opt source               destination
ACCEPT     tcp  --  anywhere             ip-172-17-0-3.ec2.internal  tcp dpt:10943
ACCEPT     tcp  --  anywhere             ip-172-17-0-7.ec2.internal  tcp dpt:11020
ACCEPT     tcp  --  anywhere             ip-172-17-0-8.ec2.internal  tcp dpt:7485
ACCEPT     tcp  --  anywhere             ip-172-17-0-8.ec2.internal  tcp dpt:7484
ACCEPT     tcp  --  anywhere             ip-172-17-0-2.ec2.internal  tcp dpt:12001
ACCEPT     tcp  --  anywhere             ip-172-17-0-2.ec2.internal  tcp dpt:12000
ACCEPT     tcp  --  anywhere             ip-172-17-0-4.ec2.internal  tcp dpt:11011
ACCEPT     tcp  --  anywhere             ip-172-17-0-4.ec2.internal  tcp dpt:11000
ACCEPT     tcp  --  anywhere             ip-172-17-0-6.ec2.internal  tcp dpt:11011
ACCEPT     tcp  --  anywhere             ip-172-17-0-6.ec2.internal  tcp dpt:11000
ACCEPT     tcp  --  anywhere             ip-172-17-0-5.ec2.internal  tcp dpt:11011
ACCEPT     tcp  --  anywhere             ip-172-17-0-5.ec2.internal  tcp dpt:11000

Chain DOCKER-ISOLATION (1 references)
target     prot opt source               destination
RETURN     all  --  anywhere             anywhere
```

**Comment by brndnmtthws** [comment]
That's odd. See if you can delete that rule:

```
# iptables -w -D INPUT -p tcp --dport 9090 --syn -j DROP
```

Someone else reported a similar problem (firewall rules getting stuck), but I'm not sure what would cause that, or why it wouldn't reset after the initial reload.

**Comment by moricard** [comment]
Deleting the rule worked fine and I can start the load balancer without problems now, thanks for the help.
I don't know if that helps but here are some of the logs when starting the marathon-lb docker container:

```
[... lots of stuff about configuring the services everything looks normal]

marathon_lb: moving temp file /tmp/tmp9cvy0kc1 to /marathon-lb/haproxy.cfg
marathon_lb: reloading using sv reload /marathon-lb/service/haproxy
ok: run: /marathon-lb/service/haproxy: (pid 16) 0s
Reloading haproxy
Configuration file is valid
cat: /tmp/haproxy.pid: No such file or directory

[WARNING] 119/154920 (44) : Can't read first line of the server state file '/var/state/haproxy/global'
15 times this line

[ALERT] 119/154920 (44) : sendmsg logger #1 failed: No such file or directory (errno=2)
[ALERT] 119/154920 (44) : sendmsg logger #2 failed: No such file or directory (errno=2)
32 times these two lines

marathon_lb: reload finished, took 0.5161159038543701 seconds
marathon_lb: updating tasks finished, took 0.5473852157592773 seconds
```

**Comment by brndnmtthws** [comment]
Those warnings can safely be ignored.

Is there any chance you changed the `PORTS` environment variable between runs? That's something that could certainly cause this behaviour.

**Comment by moricard** [comment]
I did not change the `PORTS` environment variable between runs.

**Comment by karthik2884** [comment]
We ran into the same issue. I believe if a marathon-lb instance gets killed after adding rules but before removing them after haproxy conf reload, the iptable rules remain and never get removed. When marathon-lb is restarted on the same machine it proceeds as normal but no requests reach this instance due to the iptable rules that still exist. My guess is that sometimes Marathon itself might restart a marathon-lb instance when reload takes more time than expected causing health check failures. This is particularly prominent when you have marathon-lb running on all agents (for a given agent group).

We temporarily resolved this by modifying the run script to removing iptable rules for all ports in PORTS env at the beginning.

**Comment by brndnmtthws** [comment]
Can you test out #150? Just try running MLB with the `remove-iptables-rules-at-start` tag.

**Comment by karthik2884** [comment]
@moricard were you able to test @brndnmtthws fix? Our temp fix was very similar to #150 . So would work for our issue (although did not get a chance to test it with pr yet)

**Comment by moricard** [comment]
@karthik2884 didn't have time to do so yet, I will let you know as soon as I do this.

**Comment by brndnmtthws** [comment]
I went ahead and merged #150. Please reopen the issue if this doesn't solve the problem for you.
